doc/api/quic.md

```diff
@@ -0,0 +1,1344 @@
+# QUIC
+
+## Overview
+
+The `net/quic` module provides client and server implementations of the QUIC
+and HTTP/3 protocols.
+
+```mjs
+import {
+  Endpoint,
+} from 'node:net/quic';
+```
+
+```cjs
+const {
+  Endpoint,
+} = require('net/quic');
+```
+
+The {quic.Endpoint} object is the primary entrypoint for creating
+both clients and servers.
+
+### A simple QUIC Server
+
+```mjs
+import {
+  Endpoint,
+} from 'node:net/quic';
+
+const server = new Endpoint();
+
+server.onsession = ({ session }) => {
+  session.onstream = ({ stream, respondWith }) => {
+    respondWith({ body: 'Hello World' });
+  };
+};
+
+server.listen({ alpn: 'xyz' });
+```
+
+```cjs
+const {
+  Endpoint,
+} = require('net/quic');
+
+const server = new Endpoint();
+
+server.onsession = ({ session }) => {
+  session.onstream = ({ stream, respondWith }) => {
+    respondWith({ body: 'Hello World' });
+  };
+};
+
+server.listen({ alpn: 'xyz' });
+```
+
+### A simple QUIC Client
+
+```mjs
+import {
+  Endpoint,
+} from 'node:net/quic';
+
+const client = new Endpoint();
+const session = client.connect({
+  address: '123.123.123.123',
+  port: 1234,
+}, { alpn: 'xyz' });
+const stream = session.open({ body: 'Hello World' });
+console.log(await stream.text());
+```
+
+```cjs
+const {
+  Endpoint,
+} = require('net/quic');
+
+async function sayHi() {
+  const client = new Endpoint();
+  const session = client.connect({
+    address: '123.123.123.123',
+    port: 1234,
+  }, { alpn: 'xyz' });
+  const stream = session.open({ body: 'Hello World' });
+  console.log(await stream.text());
+}
+
+sayHi();
+```
+
+### HTTP/3
+
+#### A simple HTTP/3 Server
+
+```mjs
+import {
+  Endpoint,
+} from 'node:net/quic';
+
+import { stdout } from 'node:process';
+
+const endpoint = new Endpoint({ address: { port: 12345 } });
+
+endpoint.onsession = ({ session }) => {
+  session.onstream = async ({ stream, respondWith }) => {
+    respondWith({
+      headers: { 'content-type': 'text/plain' },
+      body: 'right back at you',
+    });
+
+    console.log(await stream.headers);
+
+    stream.readableNodeStream().pipe(stdout);
+  };
```

Comment:
Have you considered something like this?

```js
session.onstream = async ({ headers, body }) => {
    console.log(headers);
    
    body.readableNodeStream().pipe(stdout);

    return {
      headers: { 
        'content-type': 'text/plain',
        ':status': 200
      },
      body: 'right back at you',
    }
  };
```

Reply:
The `onstream` call here is just a wrapper for the `EventTarget` `addEventListener()` method, returning a value here is not really possible and `EventTarget` does not know how to handle async handler functions beyond the typical default unhandled rejection handling. Also keep in mind that while this was intentionally made similar to the `fetch` event in Service Workers, this API is being designed for QUIC generically, not HTTP specifically. Headers are not going to be supported by most QUIC applications used at this level (they are only used here when we know the underlying application support supports them and node.js' internals have support for it... otherwise it will be up to application layer code to figure that out. One could build your suggested API as a layer on top of what there is currently but I don't think this suggestion works generically.